resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java

```diff
@@ -0,0 +1,289 @@
+    private static final AtomicBoolean WARN_BUF = new AtomicBoolean();
+    public static final String REQUEST_TIMEOUT_MS = JettyClientEngine.class + "$RequestTimeout";
+    // Yeah, this is the Jersey one, but there's no standard one and it makes more sense to reuse than make our own...
+    public static final String FOLLOW_REDIRECTS = "jersey.config.client.followRedirects";
+
+    private static final InvocationCallback<ClientResponse> NOP = new InvocationCallback<ClientResponse>() {
+        @Override
+        public void completed(ClientResponse response) {
+        }
+
+        @Override
+        public void failed(Throwable throwable) {
+        }
+    };
+
+    private final HttpClient client;
+
+    public JettyClientEngine(HttpClient client) {
+        if (!client.isStarted()) {
+            try {
+                client.start();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        this.client = client;
+    }
+
+    @Override
+    public SSLContext getSslContext() {
+        return client.getSslContextFactory().getSslContext();
+    }
+
+    @Override
+    public HostnameVerifier getHostnameVerifier() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ClientResponse invoke(ClientInvocation invocation) {
+        Future<ClientResponse> future = submit(invocation, true, NOP, null);
+        try {
+            return future.get(1, TimeUnit.HOURS); // There's already an idle and connect timeout, do we need one here?
+        } catch (InterruptedException e) {
+            future.cancel(true);
+            Thread.currentThread().interrupt();
+            throw clientException(e, null);
+        } catch (TimeoutException | ExecutionException e) {
+            future.cancel(true);
+            throw clientException(e.getCause(), null);
+        }
+    }
+
+    /**
+     * Implementation note: due to lack of asynchronous message decoders the request must either be buffered,
+     * or it must have a {@code null} extractor and type parameter {@code <T>} must be {@link ClientResponse},
+     * which will read the data through its stream.  It is not possible to use the synchronous JAX-RS message
+     * decoding infrastructure without buffering or spinning up auxiliary threads (arguably more expensive than buffering).
+     *
+     * @see AsyncClientHttpEngine#submit(ClientInvocation, boolean, InvocationCallback, org.jboss.resteasy.client.jaxrs.AsyncClientHttpEngine.ResultExtractor)
+     */
+    @Override
+    public <T> Future<T> submit(ClientInvocation invocation, boolean bufIn, InvocationCallback<T> callback, ResultExtractor<T> extractor) {
+        final boolean buffered;
+        if (!bufIn && extractor != null) {
+            if (!WARN_BUF.getAndSet(true)) {
+                System.err.println("TODO: ResultExtractor is synchronous and may not be used without buffering - forcing buffer mode.");
```

Comment:
@asoldano 
Logging is not used here ... but I know that some community users may want to use their custom logging instead of JBoss logging ...

Btw, what about plans of this todo?

Reply:
Yes, this could be turned into a proper log, or even removed.

The reason for this is that currently there is no asynchronous interface to `ResultExtractor`.  This means it is not possible to extract the content without either buffering, or spinning an auxiliary thread.

The solution here is to have an asynchronous interface to ResultExtractor.

Another solution is to use a coroutine technology like Quasar or Project Loom to allow suspending of the synchronous extractor.  (We have demonstrated this with Jackson, but I don't think the approach is Production Ready yet.)